src/layout.rs

```diff
@@ -366,161 +366,147 @@ impl Layout {
+                    element.size() | LE(STRONG) | f64::from(m),
+                    element.size() | EQ(WEAK) | f64::from(m),
+                ])?;
+            }
+            Constraint::Min(m) => {
+                solver.add_constraints(&[
+                    element.size() | GE(STRONG) | f64::from(m),
+                    element.size() | EQ(WEAK) | f64::from(m),
+                ])?;
             }
         }
     }
-    solver.add_constraints(&ccs).unwrap();
+
     let changes: HashMap<Variable, f64> = solver.fetch_changes().iter().copied().collect();
+
+    // convert to Rects
     let mut results = elements
         .iter()
-        .map(|element| Rect {
-            x: changes.get(&element.x).map(|&v| v as u16).unwrap_or(0),
-            y: changes.get(&element.y).map(|&v| v as u16).unwrap_or(0),
-            width: changes.get(&element.width).map(|&v| v as u16).unwrap_or(0),
-            height: changes.get(&element.height).map(|&v| v as u16).unwrap_or(0),
+        .map(|element| {
+            let start = *changes.get(&element.start).unwrap_or(&0.0);
+            let end = *changes.get(&element.end).unwrap_or(&0.0);
+            match layout.direction {
+                Direction::Horizontal => Rect {
+                    x: start as u16,
+                    y: inner.y,
+                    width: (end - start) as u16,
+                    height: inner.height,
+                },
+                Direction::Vertical => Rect {
+                    x: inner.x,
+                    y: start as u16,
+                    width: inner.width,
+                    height: (end - start) as u16,
+                },
+            }
         })
         .collect::<Rc<[Rect]>>();
 
     if layout.expand_to_fill {
-        // Fix imprecision by extending the last item a bit if necessary
-        // "unwrap" is safe, because the Rc at this point has no shared references
+        // Because it's not always possible to satisfy all constraints, the last item might be
+        // slightly smaller than the available space. E.g. when the available space is 10, and the
+        // constraints are [Length(5), Max(4)], the last item will be 4 wide instead of 5. Fix this
+        // by extending the last item a bit if necessary. (`unwrap()` is safe here, because results
+        // has no shared references right now).
         if let Some(last) = Rc::get_mut(&mut results).unwrap().last_mut() {
             match layout.direction {
-                Direction::Vertical => {
-                    last.height = dest_area.bottom() - last.y;
-                }
                 Direction::Horizontal => {
-                    last.width = dest_area.right() - last.x;
+                    last.width = inner.right() - last.x;
+                }
+                Direction::Vertical => {
+                    last.height = inner.bottom() - last.y;
                 }
             }
         }
     }
 
-    results
+    Ok(results)
 }
```

Comment:
I guess the rounding issue could result in this code being used. I'll resolve this conversation.

Reply:
This issue is triggered by the rounding problem. E.g. splitting width 2 into position = [0, 0.99], width = [0.99, 1.01] when taking the floor becomes position [0, 0], width [0, 1], which then gets modified to widths: [0, 2].

My guess is that it will probably go away with that fix.